### PR TITLE
[Kyuubi #1444] Implement api: /${version}/sessions/${session_identifier}/operations/${operation_identifier}

### PR DESCRIPTION
**Pull request by simon824**
### _Why are the changes needed?_
This is a subtask of umbrella issue #KPIP-1

/${version}/sessions/${session_identifier}/operations/${operation_identifier}
- desc: get an operation with a given session identifier and operation identifier
- method: GET
- params: none
- returns: an instance of OperationHandle

/${version}/sessions/${session_identifier}/operations/${operation_identifier}
- desc: remove operation based on a given session identifier and operation identifier
- method: DELETE
- params: none
- returns: an instance of OperationHandle

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by simon824** [comment]
cc @yanghua @ulysses-you @yaooqinn

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 1]
```diff
@@ -322,6 +322,70 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+      .isHasOperationHandle(operationHandle)
+    if (hasHandle) {
+      operationHandle
+    } else {
+      throw new NotFoundException(s"Error getting operationHandle")
+    }
+
+  }
+
+  def getOperationHandle(operationHandleStr: String): OperationHandle = {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:369]: parseOperationHandle

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 2]
```diff
@@ -322,6 +322,70 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+    try {
+      val splitOperationHandle = operationHandleStr.split("\\|")
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:371]: splitOperationHandle - > operationHandleParts

require(splitOperationHandle === 3, "some msg")

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 3]
```diff
@@ -322,6 +322,70 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+      val handleIdentifier = new HandleIdentifier(
+        UUID.fromString(splitOperationHandle(0)),
+        UUID.fromString(splitOperationHandle(1)))
+      val protocolVersion = TProtocolVersion.findByValue(splitOperationHandle(2).toInt)
+      val operationType = OperationType.withName(splitOperationHandle(3))
+      val operationHandle = new OperationHandle(handleIdentifier, operationType, protocolVersion)
+
+      // if the operationHandle is invalid, KyuubiSQLException will be thrown here.
+      backendService.sessionManager.operationManager.getOperation(operationHandle)
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:380]: we can make it method only for parsing

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 4]
```diff
@@ -322,6 +322,70 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+    }
+  }
+
+  @ApiResponse(
+    responseCode = "200",
+    content = Array(new Content(
+      mediaType = MediaType.APPLICATION_JSON)),
+    description =
+      "Get an operation handle with a given session identifier and operation identifier")
+  @GET
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:352]: I don't get the meaning of this API, why shall we use an operation handle to get itself?

**Comment by yanghua** [comment]
sorry for the late reply, will have a check soon.

**Review by yanghua — COMMENTED** [review]
@simon824 left some comments.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala (hunk 4 above)
Comment (simon824) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:352]: Is the `operation status` here duplicated with the api below?
image

Comment (yanghua) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:352]: Yeah, Maybe we can remove this API that you put here. This API is for supporting a pure Hive RPC service interface.

**Comment by simon824** [comment]
cc @yanghua @yaooqinn

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala [hunk 5]
```diff
@@ -322,7 +322,77 @@ private[v1] class SessionsResource extends ApiRequestContext {
...
+
+      val handleIdentifier = new HandleIdentifier(
+        UUID.fromString(operationHandleParts(0)),
+        UUID.fromString(operationHandleParts(1)))
+
+      val protocolVersion = TProtocolVersion.findByValue(operationHandleParts(2).toInt)
+      val operationType = OperationType.withName(operationHandleParts(3))
+      val operationHandle = new OperationHandle(handleIdentifier, operationType, protocolVersion)
+
+      if (!backendService.sessionManager.operationManager.isHasOperation(operationHandle)) {
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/api/v1/SessionsResource.scala:382]: we don't need this

**Review by yaooqinn — APPROVED** [review]

**Comment by yanghua** [comment]
Will have a final check soon.

**Review by yanghua — APPROVED** [review]
LGTM

**Comment by yanghua** [comment]
@simon824 BTW, there are some APIs that provide the capacity for `Operation`. IMO, we should introduce an `OperationResource` to place them (refer to #917 ). WDYT?

**Comment by simon824** [comment]
> @simon824 BTW, there are some APIs that provide the capacity for `Operation`. IMO, we should introduce an `OperationResource` to place them (refer to #917 ). WDYT?

Yes, I will introduce an OperationResource next PR.

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**